x-pack/metricbeat/module/mssql/log/log.go

```diff
@@ -0,0 +1,223 @@
+	backupTime            string
+	sinceLastBackupMB     *float64
+	sinceLastCheckpointMB float64
+	recoverySizeMB        float64
+}
+
+func init() {
+	mb.Registry.MustAddMetricSet("mssql", "log", New,
+		mb.DefaultMetricSet(),
+		mb.WithHostParser(mssql.HostParser))
+}
+
+// MetricSet type defines all fields of the MetricSet
+type MetricSet struct {
+	mb.BaseMetricSet
+	log *logp.Logger
+	db  *sql.DB
+}
+
+// New create a new instance of the MetricSet
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The mssql log metricset is beta.")
+
+	logger := logp.NewLogger("mssql.log").With("host", base.HostData().SanitizedURI)
+
+	db, err := mssql.NewConnection(base.HostData().URI)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create connection to db")
+	}
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		log:           logger,
+		db:            db,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right format
+// It returns the event which is then forward to the output. In case of an error, a
+// descriptive error must be returned.
+func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
+	dbs, err := m.getDbsNames()
+	if err != nil {
+		reporter.Error(err)
+		return
+	}
+
+	for _, db := range dbs {
+		moduleFields := common.MapStr{
+			"database": common.MapStr{
+				"id":   db.id,
+				"name": db.name,
+			},
+		}
+		metricsetFields := common.MapStr{}
+
+		spaceUsage, err := m.getLogSpaceUsageForDb(db.name)
+		if err != nil {
+			reporter.Error(err)
+		} else {
+			metricsetFields["space_usage"] = spaceUsage
+		}
+
+		stats, err := m.getLogStats(db)
+		if err != nil {
+			reporter.Error(err)
+		} else {
+			metricsetFields["stats"] = stats
+		}
+
+		// Both log space and log size are available, so report both
+		if isReported := reporter.Event(mb.Event{
+			ModuleFields:    moduleFields,
+			MetricSetFields: metricsetFields,
+		}); !isReported {
+			m.log.Debug("event not reported")
+		}
+
+	}
+}
+
+// Close the connection to the server at the engine level
+func (m *MetricSet) Close() error {
+	return m.db.Close()
+}
+
+func (m *MetricSet) getLogSpaceUsageForDb(dbName string) (result common.MapStr, err error) {
+
+	rows, err := m.db.Query(fmt.Sprintf(`USE %s; SELECT * FROM sys.dm_db_log_space_usage;`, dbName))
+	if err != nil {
+		return
+	}
+	defer closeRows(rows)
+
+	for rows.Next() {
+		var res logSpace
+		if err = rows.Scan(&res.id, &res.totalLogSizeInBytes, &res.usedLogSpaceInBytes, &res.usedLogSpaceInPercent,
+			&res.logSpaceInBytesSinceLastBackup); err != nil {
+			return
+		}
+
+		result = common.MapStr{
+			"total": common.MapStr{
+				"bytes": res.totalLogSizeInBytes,
+			},
+			"used": common.MapStr{
+				"bytes": res.usedLogSpaceInBytes,
+				"pct":   res.usedLogSpaceInPercent,
+			},
+			"since_last_backup": common.MapStr{
+				"bytes": res.logSpaceInBytesSinceLastBackup,
+			},
+		}
+		// According to docs "All transaction log files are combined." Means,
```

Comment:
What is this related to?

Reply:
Same here. A message for the "future me" (or any of us). By just looking at the code it might look that the result is being overriden in every scan but, in fact, it just returns a single result.

Suggestions welcome :)

Reply:
I think I get what you commented here more then what is in the comment itself. Also the location is a bit odd. I normally expect the comment before the code line that it explains.

Reply:
No. I would expect the comment on line 134 to be something like:

```
This query only returns a single result so an error in the first scan is a "data returned".
```

Reply:
Now I'm using `QueryRow` but I have still placed some comment to clarify why